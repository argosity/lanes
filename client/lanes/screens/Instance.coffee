Lanes.renderScreenTo = (selector, options)->
    return new Lanes.Screens.Instance(selector, options);


class Lanes.Screens.Instance

    constructor: (selector, options)->
        this.viewport = new Lanes.Views.Viewport({ selector: selector, instance: this });
        Lanes.Data.Bootstrap.initialize(options);
        Lanes.$(document).ready => @boot(options)

    boot: (options)->
        this.root = Lanes.$( this.viewport.selector );
        this.root.data().workspace = this;
        this.viewport.root = this.root;

        Lanes.lib.ResizeSensor(this.root[0],  _.bind( _.debounce( =>
            @viewport.set({ width: this.root.width(), height: this.root.height() });
        , 250 ), this) );

        this.root.addClass('lanes root');
        this.root.tooltip({
            viewport: '.lanes'
            selector: '[data-tooltip-message]'
            title: => @getAttribute('data-tooltip-message')
        });
        Lanes.Views.Keys.initialize();
        Lanes.Data.PubSub.initialize() if options.pub_sub;
        
        view = Lanes.getPath(options.root_view);
        if view
            this.displayInitialView(view);
        else
            definition=Lanes.Data.Screens.all.findWhere({view: options.root_view});
            if definition
                definition.getScreen().then( (screen)=>
                    # break out of the promise so an errors during render get thrown properly
                    _.defer( => @displayInitialView(screen); )
                ,(msg)->
                    Lanes.fatal("Unable to load initial screen ${options.root_view}", msg);
                )
            else
                Lanes.fatal(options.root_view + " doesn't exist!");

    displayInitialView:(view)->
        this.view = new view({parent: this, model: this.viewport}).render();
        this.viewport.el = this.view.$el;

        this.root.append( this.view.el );
        Lanes.Extensions.fireOnAvailable(this);


