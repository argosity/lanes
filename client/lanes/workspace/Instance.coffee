Lanes.Workspace.create = (selector, options, extension_data)->
    return new Lanes.Workspace.Instance(selector, options, extension_data);


class Lanes.Workspace.Instance

    constructor: (selector, options)->
        this.viewport = new Lanes.Views.Viewport({ selector: selector, instance: this });
        Lanes.Data.Bootstrap.initialize({
            csrf: options.csrf_token,
            root: options.api_path,
            data: options
        });
        Lanes.$(document).ready => @boot(options)

    boot: (options)->
        this.root = Lanes.$( this.viewport.selector );
        this.root.data().workspace = this;
        this.viewport.root = this.root;

        Lanes.lib.ResizeSensor(this.root[0],  _.bind( _.debounce( =>
            @viewport.set({ width: this.root.width(), height: this.root.height() });
        , 250 ), this) );

        this.root.addClass('lanes root');
        this.root.tooltip({
            viewport: '.lanes'
            selector: '[data-tooltip-message]'
            title: => @getAttribute('data-tooltip-message')
        });
        Lanes.Views.Keys.initialize();
        Lanes.Data.PubSub.initialize() if options.pub_sub;
        
        view = Lanes.getPath(options.root_view);
        if view
            this.displayInitialView(view);
        else
            definition=Lanes.Data.Screens.all.findWhere({view: options.root_view});
            if definition
                definition.getScreen().then( (screen)=>
                    # break out of the promise so an errors during render get thrown properly
                    _.defer( => @displayInitialView(screen); )
                ,(msg)->
                    Lanes.fatal("Unable to load initial screen ${options.root_view}", msg);
                )
            else
                Lanes.fatal(options.root_view + " doesn't exist!");




    displayInitialView:(view)->
        this.view = new view({parent: this, model: this.viewport}).render();
        this.viewport.el = this.view.$el;

        this.root.append( this.view.el );
        Lanes.Extensions.fireOnAvailable(this);


