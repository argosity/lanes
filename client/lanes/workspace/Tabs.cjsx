class TabView extends Lanes.React.Component

    activate: ->
        @model.active = true
        null

    close: ->
        @model.remove()

    render: ->
        <li key={@model.cid} className={_.classnames(active: @model.active)}>
            <a onClick={@activate} className='tab'>{@model.screen.title}</a>
            <span onClick={@close} className='close'>Ã—</span>
        </li>


class Lanes.Workspace.Tabs extends Lanes.React.Component

    dataObjects:
        collection: -> Lanes.Screens.Definitions.displaying

    render: ->
        <div className="menu-container">
            <div className="active-screens">
                <div className="scroller scroller-left">
                    <i className="glyphicon glyphicon-chevron-left"/>
                </div>
                <div className="scroller scroller-right">
                    <i className="glyphicon glyphicon-chevron-right"/>
                </div>
                <div className="wrapper">
                    <ul className="nav nav-tabs">
                        { @collection.map (view) ->
                            <TabView key={view.id} model=view /> }
                    </ul>
                </div>
            </div>
        </div>
