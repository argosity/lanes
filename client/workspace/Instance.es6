Lanes.Workspace.create = function(selector, options, extension_data){
    return new Lanes.Workspace.Instance(selector, options, extension_data);
};

Lanes.Workspace.Instance = class Instance {

    constructor(selector, options, extension_data){
        _.extend(this,options);

        this.viewport = new Lanes.Views.Viewport({ selector: selector, instance: this });
        Lanes.Data.Bootstrap.initialize({
            csrf: options.csrf_token,
            root: options.root_path,
            data: extension_data
        });
        Lanes.$(document).ready( ()=> {
            this.boot();
        });
    }



    boot(){
        this.root = Lanes.$( this.ui.selector );
        this.root.data().workspace = this;
        this.ui.root = this.root;

        Lanes.lib.ResizeSensor(this.root[0],  _.bind( _.debounce( function(a,b,c){
            this.ui.set({ viewport_width: this.root.width(), viewport_height: this.root.height() });
        }, 250 ), this) );

        this.root.addClass('lanes root');
        this.root.tooltip({
            viewport: '.lanes',
            selector: '[data-tooltip-message]',
            title: function(){ return this.getAttribute('data-tooltip-message'); }
        });

        var view = Lanes.getPath(this.view);
        if (!view){
            Lanes.fatal(this.view + " doesn't exist!");
        }
        this.view = new view({parent: this, model: this.ui}).render();
        this.ui.viewport = this.view.$el;

        this.root.append( this.view.el );

        Lanes.Views.Keys.initialize();
        Lanes.Data.PubSub.initialize();
        Lanes.Extensions.fireOnAvailable(this);

        this.login_dialog = new Lanes.Views.LoginDialog({ parent: this });

        // var screen = Lanes.Data.Screens.all.findWhere({ id: 'user-maintenance' });
        // screen.display();
    }
};
