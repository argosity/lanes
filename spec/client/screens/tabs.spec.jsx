import React from 'react';
import { each, range } from 'lodash';

import Tabs, { ScreenTab } from 'lanes/workspace/tabs';
import Instance    from 'lanes/screens/instance';

import { getTestScreen } from '../test-models';
import { Context, Snapshot } from '../test-utils';

const addTab = () =>
    new Instance({ definition: getTestScreen() });

describe('Workspace Navbar', () => {
    beforeEach(() => {
        each(range(0, 3), addTab);
    });

    it('renders and adds/removes tabs', () => {
        const options = new Context();
        const tabs = shallow(<Tabs />, options);
        expect(tabs.find('ScreenTab').length).toEqual(3);
        addTab();
        expect(tabs.find('ScreenTab').length).toEqual(4);
        Instance.displaying[2].remove();
        expect(tabs.find('ScreenTab').length).toEqual(3);
        expect(Snapshot(<Tabs />)).toMatchSnapshot();
    });

    it('renders a tab', () => {
        const options = new Context();
        const screen = Instance.displaying[2];
        const tab = shallow(<ScreenTab screen={screen} />, options);
        expect(tab).toHaveRendered(`Tab[title="${screen.title}"]`);
        expect(Snapshot(<ScreenTab screen={Instance.displaying[2]} />)).toMatchSnapshot();
    });
});
