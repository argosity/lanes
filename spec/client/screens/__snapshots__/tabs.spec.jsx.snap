// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace Navbar renders a tab 1`] = `
<li
  className="grommetux-tab"
  id={undefined}
>
  <button
    aria-expanded={undefined}
    aria-label={undefined}
    aria-selected={undefined}
    className="grommetux-button grommetux-button--plain grommetux-tab__button"
    disabled={false}
    href={undefined}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="tab"
    type="button"
  >
    <label
      className="grommetux-tab__label"
      htmlFor={undefined}
    >
      title
    </label>
  </button>
</li>
`;

exports[`Workspace Navbar renders and adds/removes tabs 1`] = `
<div
  role="tablist"
>
  <ul
    className="grommetux-tabs grommetux-tabs--justify-start grommetux-tabs--responsive"
  >
    <li
      className="grommetux-tab"
      id={undefined}
    >
      <button
        aria-expanded={undefined}
        aria-label={undefined}
        aria-selected={undefined}
        className="grommetux-button grommetux-button--plain grommetux-tab__button"
        disabled={false}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="tab"
        type="button"
      >
        <label
          className="grommetux-tab__label"
          htmlFor={undefined}
        >
          title
        </label>
      </button>
    </li>
    <li
      className="grommetux-tab"
      id={undefined}
    >
      <button
        aria-expanded={undefined}
        aria-label={undefined}
        aria-selected={undefined}
        className="grommetux-button grommetux-button--plain grommetux-tab__button"
        disabled={false}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="tab"
        type="button"
      >
        <label
          className="grommetux-tab__label"
          htmlFor={undefined}
        >
          title
        </label>
      </button>
    </li>
    <li
      className="grommetux-tab"
      id={undefined}
    >
      <button
        aria-expanded={undefined}
        aria-label={undefined}
        aria-selected={undefined}
        className="grommetux-button grommetux-button--plain grommetux-tab__button"
        disabled={false}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="tab"
        type="button"
      >
        <label
          className="grommetux-tab__label"
          htmlFor={undefined}
        >
          title
        </label>
      </button>
    </li>
  </ul>
  <div
    aria-label="Tab Contents"
    role="tabpanel"
  />
</div>
`;
